src/ore/src/bytes.rs

```diff
@@ -311,36 +319,43 @@ impl<const N: usize> FromIterator<Vec<u8>> for SegmentedBytes<N> {
 }
 
 mod internal {
-    use std::collections::BTreeMap;
     use std::io;
-    use std::ops::Bound;
 
-    use crate::bytes::MaybeLgBytes;
+    use smallvec::SmallVec;
+
+    use crate::bytes::{MaybeLgBytes, SegmentedBytes};
     use crate::cast::CastFrom;
 
     /// Provides efficient reading and seeking across a collection of segmented bytes.
     #[derive(Debug)]
-    pub struct SegmentedReader {
-        segments: BTreeMap<usize, MaybeLgBytes>,
+    pub struct SegmentedReader<const N: usize = 1> {
+        segments: SmallVec<[(MaybeLgBytes, usize); N]>,
+        /// Total length of all segments.
         len: usize,
-        pointer: u64,
+        // Overall byte position we're currently pointing at.
+        overall_ptr: usize,
+        /// Current segement that we'd read from.
+        segment_ptr: usize,
     }
 
-    impl SegmentedReader {
-        pub fn new(segments: impl IntoIterator<Item = MaybeLgBytes>) -> Self {
-            let mut map = BTreeMap::new();
-            let mut total_len = 0;
+    impl<const N: usize> SegmentedReader<N> {
+        pub fn new(mut bytes: SegmentedBytes<N>) -> Self {
+            // Filter out empty segments.
+            let mut accum_length = 0;
+            bytes.segments.retain(|(segment, len)| {
+                // Re-adjust our accumlated lengths.
+                accum_length += segment.len();
+                *len = accum_length;
```

Comment:
Wondering if we could shift some of this work earlier in the flow?
- Initialize `len` to the correct value in `SegmentedBytes` instead of fixing it up here. (`SegmentedBytes` already seems to keep the running length.)
- Drop empty segments at `push` time.

Reply:
Initializing `len` to the correct values in `SegmentedBytes` is tricky because of the `bytes::Buf::advance` impl that drops segments and thus would invalidate the lengths. It's not impossible to fix, just need to spend some time thinking about the best way to do it. Realistically any code path that needs `bytes::Buf` won't need `std::io::Read` and vice versa, so maybe type wrappers is the best way and we can avoid any runtime computation?

I like the idea of dropping empty segments at `push` time, I made that change!

Reply:
> It's not impossible to fix, just need to spend some time thinking about the best way to do it.

One possibility is to track an `advanced_by` field explicitly. (And then the `SegmentedBytes::len` is something like `segments.last().map_or(0, |(_, n) n - advanced_by` or whatever.)

Anyways, not critical if it's a hassle - I just find stuff like this tends to be less bug-prone if state doesn't change its meaning in different contexts, and this seemed relatively cheap to make meaningful.

> I like the idea of dropping empty segments at push time, I made that change!

For context: does that mean this `retain` should never reject anything?

Reply:
> Anyways, not critical if it's a hassle - I just find stuff like this tends to be less bug-prone if state doesn't change its meaning in different contexts, and this seemed relatively cheap to make meaningful.

Totally! I'll think on this a bit more. I like the approach with `advanced_by` the only complication is we would need to pass it down into `SegmentedReader` as well since we shouldn't ever be able to seek before that point.

> For context: does that mean this `retain` should never reject anything?

Exactly! I actually removed the `retain` and changed it to a loop that iterates over mutable references

Reply:
Messed around with this a bit and it was a bit tricky to get right, going to punt on it for now